### ISSUE TITLE
Add support for new Summon APIs

### ISSUE THREAD

**Issue by timdorr** [timdorr/model-s-api#16]
You can now "Summon" via the app and there is also some HomeLink capability. I have a 2012 Model S, so I'm unable to live test this stuff, but hopefully I can use the HomeLink APIs nonetheless.

**Comment by btamburrino** [comment]
@timdorr I was updating a PHP library with the new info from your code and I ran into a stumbling block. While data_requests were working, anything that was a command (or wake_up) was returning a 404 error. Is it because of the Client ID and Secret that I'm using, which I got from your page? I'm able to do the commands from the native apps.

It looks like folks on TMC found the endpoints for the 7.1 features, but since I can't even honk my horn right now I haven't bothered to try them.

**Comment by timdorr** [comment]
@btamburrino Make sure you're issuing POST requests for the `command` endpoints. They are not GETs.

**Comment by btamburrino** [comment]
@timdoor I tried that, they still returned 404's.

One thing I can report is that get all vehicles now returns `id_s` which is a stringified version of id. This is good because the id is now a BIGINT and that messes with some JSON decoders (like PHP's.)

**Comment by dcdspace** [comment]
Did you figure out what the endpoints are for summon and homelink? I can't seem to find them anywhere on the web...

**Comment by timdorr** [comment]
I mentioned them here: http://www.teslamotorsclub.com/showthread.php/13410-Model-S-REST-API/page49?p=1337610&viewfull=1#post1337610

They both require keyfob presence in the car, so it's hard to test. I haven't dragged my laptop out to the garage yet.

**Comment by timdorr** [comment]
Just found out you can trick it by doing a keyless start and then using the associated APIs. I'll test this over the weekend. But I get the feeling Tesla is going to close up that hole pretty quickly (possibly even through the API itself...)

**Comment by dcdspace** [comment]
Thanks I'll have to test it out tomorrow. Also you said "It looks like token is just your bluetooth device name", so is this token for the https://owner-api.teslamotors.com/api/1/vehicles/vehicle_id/trigger_homelink different from the token we normally use that is returned from the oauth request?

Also I know that the phone has to be near the car so once you fake lat/long parameters to just be the car's current position, you're saying after doing a keyless start it will work? Pretty cool...

**Comment by timdorr** [comment]
Yes, it's a separate token field. I traced the code back to the Bluetooth device name in the Android app. 

And yeah, as long as the lat lon is within a certain distance (probably the magical 39 feet), it should work. It should be fine to simply duplicate the car's reported position.

**Comment by mseminatore** [comment]
I see my issue was closed dupe with this one.  I am unable to get homelink to work.  Have you made progress by chance?  I have a 2013 and always get **reason:"unavailable"** when I try to invoke.  I am sending the current vehicle lat/long and phone BT name.  I've tried enabling remote start first.

**Comment by timdorr** [comment]
I have a 2012, so I'm probably in the same boat. I'm pretty sure they grouo this off with Summon. I haven't made any attempts yet, though.

**Comment by dcdspace** [comment]
I don't pass the BT device name and it works well for both summon and homelink. I instead just pass the same token I've been using for everything else.

**Comment by mseminatore** [comment]
@dcdspace Do you have AP hardware?

**Comment by dcdspace** [comment]
Yes

**Comment by dcdspace** [comment]
Tesla must have updated their api because "trigger_homelink" no longer works, yet the native Tesla app can still do this. Any ideas?

**Comment by timdorr** [comment]
Are you attempting it with the fob within range? I think they make have closed the keyless driving loophole. As far as I know, the API itself hasn't changed.

**Comment by dcdspace** [comment]
Yes the fob is in the car, and when I press the home link button in the Tesla app it works but my own api request no longer works. If they didn't update their API then what changed?

Also I tried to analyze their app with a proxy tool like Charles but the connection fails when I try that, any advice?

--
Dylan Diamond

**Comment by dcdspace** [comment]
@timdorr here is the explanation: https://teslamotorsclub.com/tmc/threads/model-s-rest-api.13410/page-80

They switched to web sockets apparently so it will be a bit harder to call summon and homelink.

**Comment by timdorr** [comment]
Looks like you open a standard Websocket connection to `wss://streaming.vn.teslamotors.com/connect/{vehicle_id}`. 

From there, there is a JSON-based communication protocol with output commands and input results from the server: It has the rough format of 
```json
// Command from client
{"msg_type":"autopark:cmd_forward", "latitude":123.456, "longitude": 78.90}

// Response from server
{"cmd_type":"autopark:cmd_forward"","reason":"", "result":true}
```
So, it's pretty analogous to other parts of the standard HTTP API. I'm looking through all the commands right now and there are a bunch. I'll add the high level here and fill in the parameters that I see.

```
control:hello
control:goodbye
control:pong
autopark:status
vehicle_data:location
autopark:cmd_result
autopark:heartbeat_car
autopark:error
homelink:status
homelink:cmd_result
control:ping
autopark:heartbeat_app
autopark:cmd_abort
autopark:cmd_forward
autopark:cmd_reverse
homelink:cmd_trigger
```

**Comment by timdorr** [comment]
Commands that can be received (`msg_type -> arg1, arg2, ...`) :
```
autopark:status -> autopark_state
autopark:cmd_result -> cmd_type, reason, result
autopark:heartbeat_car -> timestamp
autopark:error -> error_type
homelink:status -> homelink_nearby (bool)
homelink:cmd_result -> command_type, failure_reason, result
vehicle_data:location -> latitude, longitude
```

**Comment by timdorr** [comment]
Commands that can be sent (`msg_type -> arg1, arg2, ...`):
```
control:ping -> timestamp
autopark:heartbeat_app -> timestamp
autopark:cmd_abort -> 
autopark:cmd_forward -> latitude, longitude
autopark:cmd_reverse -> latitude, longitude
homelink:cmd_trigger -> latitude, longitude
```

**Comment by timdorr** [comment]
For those really interested, here's the entire state machine they use for autopark related stuff (it's pretty complex): https://gist.github.com/timdorr/61834c6ada93d961e4646ea5fff6d7f7

**Comment by dcdspace** [comment]
@timdorr Thanks so much! I just tested this and it works. A few questions I'm hoping you'll have the answers to:

I noticed that the headers that are sent with the initial websocket have a different Authorization value than normally. It is not `Bearer: {token}` but Basic with something else, so I'm wondering how to get that other value. 

Do we need to include the `Sec-WebSocket-Key` value? If so where do we retrieve it?

Here is a screenshot of the headers with private info redacted. Thanks.

![screen shot 2016-06-21 at 1 16 01 am](https://cloud.githubusercontent.com/assets/6096771/16218690/2d34485c-374e-11e6-9f51-55ab0111381e.png)

**Comment by pdb0102** [comment]
Sec-WebSocket-Key is a mechanism to prevent proxies from caching any WebSocket stream being passed through them, by altering the data every time. It should be automatically handled by any WebSocket framework used to communicate. (In short, it's a SHA1 hash that is generated by taking the previous value, hashing it with a particular guid, and then Base64 encoding that)

The Authorization: Basic is the standard HTTP auth mechanism, a base64 encoding of username, followed by a colon (:) and then the password. The username is your Tesla account email address, the password being sent is the text/hex representation of 8 bytes , but I haven't been able to determine how that's generated yet. It is not a partial md5/sha1/sha256 hash of the tesla password or bearer token, nor a part of the bearer token itself. I'm assuming they're transforming the bearer token by some means to derive the password, but nothing obvious jumped out at me.

**Comment by timdorr** [comment]
Correct. The header is in the format `Authorization: Basic %{base64(token)}`. The token comes from the vehicles endpoint in the `tokens` field. Same as the streaming data API.

**Comment by pdb0102** [comment]
Duh, yeah. Forgot about the streaming tokens. But wouldn't the format be `Authorization: Basic %{base64(email_addr ':' token)}` ?

**Comment by timdorr** [comment]
Ah, yes, you're right. It's `Basic {base64({email}:{token})}`

**Comment by dcdspace** [comment]
Thanks it works now. Also what rate should we send the control:ping at?

**Comment by pdb0102** [comment]
@dcdspace The Tesla implementation seems to send a ping every 5 seconds. The server will respond with a pong having the same time stamp as the ping you sent.

**Comment by dcdspace** [comment]
@pdb0102 Ok awesome. And when doing the autopark, should I send a heartbeat every 0.5 seconds? I have it going on a timer so that the user doesn't have to hold down a button so I'm hoping the car will stop on its own after a certain amount of time...?

**Comment by timdorr** [comment]
Look at whatever the `heartbeat_frequency` is in the `control:hello` response.

**Comment by pdb0102** [comment]
@dcdspace I'm assuming you're talking about summon automatically stopping after a certain amount of time? That's actually configured in the car, how many feet your summon goes at most before stopping by itself (assuming it doesn't encounter an obstacle first). So I would assume if you keep sending the heartbeat as requested in `control:hello` it will keep driving itself until the configured distance is reached. I'd assume that when it's done driving, either all the way, or when it encounters an obstacle, you get one or more of the `autopark:` responses from the car telling you "how it went".

**Comment by SamGabbay** [comment]
@timdorr Hey, thanks for all your hard work! I keep getting "Can't validate password." And the token I am generating is valid.

**Comment by artiestie** [comment]
I'm lost with the Basic format.  I've tried base64 encoding my {email}:{token} and all the variations.  How are you formatting the header?

**Comment by pdb0102** [comment]
@artiestie It's a standard HTTP Authentication header. See https://en.wikipedia.org/wiki/Basic_access_authentication

**Comment by artiestie** [comment]
Indeed it is.

**Comment by pdb0102** [comment]
@SamGabbay Can you provide some detail as to what you're doing/sending that results in that error?

**Comment by artiestie** [comment]
@pdb0102 We are sending request to 

wss://streaming.vn.teslamotors.com/connect/myVehicleID with header key `Authorization` and value `Basic aowjef0923jf` 

where *aowjef0923jf* is the base64-encode of "email:token"

**Comment by jsgoecke** [comment]
Yes, I am also getting a 401 when I do this:

```ruby
  encodedToken = Base64.encode64('my_email:my_access_token')
  ws = Faye::WebSocket::Client.new(
  	'wss://streaming.vn.teslamotors.com/connect/real_vehicle_id',
  	nil,
  	:headers => {
  		'Authorization' => 'Basic ' + encodedToken
  	})
```

**Comment by timdorr** [comment]
@jsgoecke Use `Base64.strict_encode64`. `encode64` adds line breaks.

**Comment by jsgoecke** [comment]
@artiestie I am having a similar issue just trying to get the basics working. This script:

```ruby
require 'faye/websocket'
require 'eventmachine'
require 'base64'

endpoint = 'wss://streaming.vn.teslamotors.com/connect/#{valid_vehicle_id}'
email = '#{valid_email}'
token = '#{valid_token}'

EM.run {
  ws = Faye::WebSocket::Client.new(
  	endpoint,
  	nil,
    {
    	:headers => {
    	 'Authorization' => "Basic " + Base64.strict_encode64(email + ':' + token)
  	 }
    }
  )
  p [:headers, ws.headers]

  ws.on :error do |event|
   p [:error, event.message, event.type]
  end

  ws.on :open do |event|
    p [:open, event.message]
  end

  ws.on :message do |event|
    p [:message, event.message]
  end

  ws.on :close do |event|
    p [:close, event.code, event.reason, event.type]
    ws = nil
  end
}
```
Results in:

```
[:headers, {}]
[:error, "Error during WebSocket handshake: Unexpected response code: 401", "error"]
[:close, 1002, "Error during WebSocket handshake: Unexpected response code: 401", "close"]
```
If I remove my Authorization header I get a 400, so the fact that headers returns an empty hash on the websocket is strange too. Not sure what I might be doing wrong here, since it looks to be the same as @timdorr is doing now in his library:

https://github.com/timdorr/model-s-api/commit/93b56434558745dd6423a2f803761ea3f4444832#diff-14d7ffa828564ca013a870a7a7b7262dR42

**Comment by jsgoecke** [comment]
I now have things working. I was using the AccessToken instead of the first Vehicle Token. Not the first time I made that mistake.

So, this works in Go now too (sloppy, but just poking):

```go
package main

import (
    "strconv"

    "github.com/jsgoecke/tesla"

    "encoding/base64"
    "log"
    "net/http"
    "net/url"
    "os"
    "os/signal"
    "time"

    "github.com/gorilla/websocket"
)

var (
    email    = "valid_email"
    server   = "streaming.vn.teslamotors.com"
    resource = "/connect/"
)

func main() {
    token, vehicleID := getTesla()
    interrupt := make(chan os.Signal, 1)
    signal.Notify(interrupt, os.Interrupt)

    u := url.URL{Scheme: "wss", Host: server, Path: resource + vehicleID}
    headers := setHeaders(token)
    log.Printf("%s", u.String())
    log.Println(headers)

    c, _, err := websocket.DefaultDialer.Dial(u.String(), headers)
    if err != nil {
        log.Fatal("dial:", err)
    }
    defer c.Close()

    done := make(chan struct{})

    go func() {
        defer c.Close()
        defer close(done)
        log.Println("launching listener")
        for {
            _, message, err := c.ReadMessage()
            if err != nil {
                log.Println("read:", err)
                return
            }
            log.Printf("recv: %s", message)
        }
    }()

    log.Println("launching watcher")
    for {
        select {
        case <-interrupt:
            log.Println("interrupt")
            select {
            case <-done:
                log.Println("done")
            case <-time.After(1 * time.Second):
            }
            c.Close()
            return
        }
    }
}

func setHeaders(token string) http.Header {
    req, _ := http.NewRequest("GET", "http://localhost", nil)
    data := []byte(email + ":" + token)
    encodedToken := base64.StdEncoding.EncodeToString(data)
    req.Header.Add("Authorization", "Basic "+encodedToken)
    return req.Header
}

func getTesla() (string, string) {
    client, err := tesla.NewClient(
        &tesla.Auth{
            ClientID:     os.Getenv("TESLA_CLIENT_ID"),
            ClientSecret: os.Getenv("TESLA_CLIENT_SECRET"),
            Email:        os.Getenv("TESLA_USERNAME"),
            Password:     os.Getenv("TESLA_PASSWORD"),
        })
    if err != nil {
        panic(err)
    }

    vehicles, err := client.Vehicles()
    if err != nil {
        panic(err)
    }
    vehicle := vehicles[0]
    return vehicle.Tokens[0], strconv.Itoa(vehicle.VehicleID)
}
```

Result:

```json
{"autopark":{"autopark_pause_timeout":2000,"autopark_stop_timeout":10000,"heartbeat_frequency":500},"connection_timeout":20000,"msg_type":"control:hello"}
```

**Comment by Swinny1989** [comment]
Hi guys,

I'm having an issue getting this to work on iOS with Objective C. I'm using [SocketRocket](https://github.com/facebook/SocketRocket) to try and open up a connection but am consistently getting a 401 error code:

`Error Domain=SRWebSocketErrorDomain Code=2132 "received bad response code from server 401" UserInfo={NSLocalizedDescription=received bad response code from server 401, HTTPResponseStatusCode=401}` 

Here is how i'm attempting to open up the connection:
```
    NSString *vehicleID = [self retrieveVehicleID];
    NSString *urlString = [NSString stringWithFormat:@"wss://@streaming.vn.teslamotors.com/connect/%@", vehicleID];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:urlString]];

    
    NSString *email = @"email_here";
    NSString *vehicleToken = @"first_vehicle_token_here";
    NSString *authString = [NSString stringWithFormat:@"%@:%@", email, vehicleToken];
    NSData *tokenData = [authString dataUsingEncoding:NSUTF8StringEncoding];
    NSString *base64Token = [tokenData base64EncodedStringWithOptions:kNilOptions];
    
    NSString *authValue = [NSString stringWithFormat:@"Basic %@", base64Token];
    [request setHTTPMethod:@"GET"];
    
    [request setValue:authValue forHTTPHeaderField:@"Authorization"];
    
    SRWebSocket *webSocket = [[SRWebSocket alloc] initWithURLRequest:[request copy]];
    webSocket.delegate = self;
    [webSocket open];
```

If anyone has any idea where i'm going wrong please let me know, the code looks correct from what I can see from the above examples yet i'm still getting these 401 errors.

Any help would be great.

**Comment by pdb0102** [comment]
@Swinny1989 I don't do iOS dev, but looking at it, is the @ after wss:// intended?

**Comment by jonasman** [comment]
that @ at the url seems wrong.
Is this your app https://itunes.apple.com/gb/app/energi-simple-remote-for-tesla/id1102475223?mt=8 ?

**Comment by Swinny1989** [comment]
@pdb0102 No it wasn't intended, that was from a test that I was trying to do the authentication inline. I have removed that again, still the same result though.

@jonasman I am the Lead Developer working on this project for @SamGabbay he's a great guy :)

**Comment by jonasman** [comment]
@Swinny1989 if you guys need a Swift lib i made one here: https://github.com/jonasman/TeslaSwift
I will change in the future some APIs and add the summon via webstocks.

**Comment by jonasman** [comment]
nice to know. sam is the guy from tesla siri :D

**Comment by Swinny1989** [comment]
@jonasman He became quite the celebrity! Even Tesla are taking note of him after he showed them his app in person lol, some employees even said it's better than their official app!

**Comment by jonasman** [comment]
@Swinny1989 did you try to "POST" not "GET"?

**Comment by Swinny1989** [comment]
@jonasman Yeah I've tried both.. i'm losing my mind here.

**Comment by timdorr** [comment]
WebSockets must be a GET.

**Comment by timdorr** [comment]
For the token, I would make sure you're getting a fresh one from the API when running this code. They are not long-lasting, so you might be testing against one that's expired.

**Comment by timdorr** [comment]
Also, the vehicle ID needs to be the `vehicle_id` field from the `/vehicles` endpoint. The `id` field is separate (probably local to API app's database) and doesn't act as a global identifier for the streaming server.

**Comment by pdb0102** [comment]
@Swinny1989 If all else fails you can kick me over a test app. I still have my setup I used to trace and decode the websocket protocol; I can run your app on my jail broken device and look what goes over the wire vs what should go over the wire.

**Comment by SamGabbay** [comment]
@timdorr First of all, thank you for all your hard work and your support! Can we all give it up for this guy? I mean he really deserves a lot of credit. 

We are using a brand new vehicle token. Still 401 error.

**Comment by morrisonbrett** [comment]
Thanks @timdorr. Your efforts are appreciated!

**Comment by timdorr** [comment]
Some possible avenues for diagnosing:

- Log out the base64 string and then decode it on the command line to make sure it's coming through right (`echo 'whateverABC123==' | base64 -d`)
- Try your inputs (vehicle ID, email, token) on another tool or sample code. My code in #31 works fine, although I don't have AP on my car to test against.
- Can SocketRocket print out debug info? If not, can you Wireshark the network traffic from the simulator (I assume that's where this is running) to see what's going over the wire? I would check for missing or mangled headers.

**Comment by SamGabbay** [comment]
@timdorr Got it working! Thank you! Will post the rest of my results very shortly. Thank you for all your help!

**Comment by Swinny1989** [comment]
@timdorr Thank you so much, got it all working now. vehicle_id was the issue

**Comment by dabbler** [comment]
@timdorr I'm trying to use your gem code.   And when I call model_s.start_autopark()
I get this:
/var/lib/gems/2.3.0/gems/tesla_api-1.3.0/lib/tesla_api/autopark.rb:15:in `block (2 levels) in start_autopark': undefined method `call' for nil:NilClass (NoMethodError)
Did you mean?  caller
	from /var/lib/gems/2.3.0/gems/websocket-driver-0.6.4/lib/websocket/driver/event_emitter.rb:39:in `block in emit'
[...snipped a bunch of intermediate lines...]
`receive_data'
	from /var/lib/gems/2.3.0/gems/eventmachine-1.2.0.1/lib/eventmachine.rb:194:in `run_machine'
	from /var/lib/gems/2.3.0/gems/eventmachine-1.2.0.1/lib/eventmachine.rb:194:in `run'
	from /var/lib/gems/2.3.0/gems/tesla_api-1.3.0/lib/tesla_api/autopark.rb:6:in `start_autopark'
	from tesla-state.rb:32:in `<main>'

I think I'm probably just not understanding how to use ruby properly.   Can you see what I'm doing wrong or do you have any sample summon code?

**Comment by timdorr** [comment]
You need to pass a block to that function. Check the code at the top of #31. Sorry, but it's undocumented for now. But try reading through the code, as it's not that complex.

**Comment by timdorr** [comment]
Bad news: [I no longer have a working Model S](https://i.redd.it/4pasx5h3eapx.jpg)

Good news: [I'm getting a new Model S that will have AP](http://i.imgur.com/IzB3Avl.png) (P90D! Woo!) 

So, soon I'll be able to test all of this crap for real!

**Comment by dabbler** [comment]
Oh my!   I do hope you (and others) are okay.

**Comment by dewski** [comment]
> Bad news: I no longer have a working Model S

Oh that was you, I saw that on Reddit. Glad you're ok!

**Comment by icdevin** [comment]
I dunno if anyone noticed but I also get a message type of `autopark:style` with one property called `autopark_style` with a value of `standard`...


```json
{ autopark_style: 'standard', msg_type: 'autopark:style' } { masked: false,
  buffer: <Buffer 7b 22 61 75 74 6f 70 61 72 6b 5f 73 74 79 6c 65 22 3a 22 73 74 61 6e 64 61 72 64 22 2c 22 6d 73 67 5f 74 79 70 65 22 3a 22 61 75 74 6f 70 61 72 6b 3a ... >,
  binary: true }
```

My car is current at the SC so I can't test much but I was just playing around and noticed nobody had mentioned it.

**Comment by timdorr** [comment]
Sounds like it's a change to handle the upcoming AP2 hardware. My car will have that hardware, so I can confirm when it gets here in a month.

**Comment by timdorr** [comment]
Just got my new car and it's online with the API. Unfortunately, as is the case with all AP2 cars for the moment, the hardware isn't active and I can't test it yet:

```
"control:hello"
{"autopark"=>{"autopark_pause_timeout"=>2000, "autopark_stop_timeout"=>10000, "heartbeat_frequency"=>500}, "connection_timeout"=>20000}
"autopark:style"
{"autopark_style"=>"dead_man"}
"autopark:status"
{"autopark_state"=>"unavailable"}
"homelink:status"
{"homelink_nearby"=>false}
```
I did hear that 8.1 is ahead of schedule and will turn most of it back on, so I'll get to testing things out then. Just have to be patient :)

**Comment by flexgames** [comment]
@SamGabbay Curious if you could post your Obj-C or Swift code on the authentication piece - you wrote : 

> @timdorr Got it working! Thank you! Will post the rest of my results very shortly. Thank you for all your help!

thanks!

**Comment by jonasman** [comment]
You can also use my swift lib:
https://github.com/jonasman/TeslaSwift

Sent from my iPhone

On 12 Jan 2017, at 02:11, flexgames <notifications@github.com<mailto:notifications@github.com>> wrote:


@SamGabbay<https://github.com/SamGabbay> Curious if you could post your Obj-C or Swift code on the authentication piece - you wrote :

@timdorr<https://github.com/timdorr> Got it working! Thank you! Will post the rest of my results very shortly. Thank you for all your help!

thanks!

-

**Comment by flexgames** [comment]
@jonasman thank you. I was looking at your lib earlier but I really need that streaming component. After researching this further it's not as easy as the modifying the endpoint POST commands to pull JSON, since it's a web socket (and not returning JSON but rather a stream of data). I'm trying to figure out the right way to poll the socket in Swift (without resorting to using a third party framework like websocket.swift, Starscream or SocketRocket if possible). I am checking the java code implementation since that's the most similar to Swift, but it requires an InputBuffer and a socket stream of which neither are really well supported in Foundation by default. 

Trying [websocket.swift](https://github.com/tidwall/SwiftWebSocket) I'm calling the ws this way:

`        let tokenEndpoint: String = "ws://streaming.vn.teslamotors.com/stream/\(car)/?values=speed,odometer"
        let url = URL(string: tokenEndpoint)
        
        //create dictionary with your parameters
        let username = "myemailhere"
        let password = token // passed in 
        let loginString = NSString(format: "%@:%@", username, password)
        let loginData: NSData = loginString.data(using: String.Encoding.utf8.rawValue)! as NSData
        let base64LoginString = loginData.base64EncodedString(options: NSData.Base64EncodingOptions())
        
        var request = URLRequest(url: url!)
        request.httpMethod = "GET"
        request.addValue("Basic \(base64LoginString)", forHTTPHeaderField: "Authorization")
        let ws = WebSocket(request:request)
        ws.event.open = {
            print("opened")
        }
        
        ws.event.close = { code, reason, clean in
            print("close")
        }
        
        ws.event.error = { error in
            print("error \(error)")
        }
        
        ws.event.message = { message in
            if let text = message as? String {
                print("recv: \(text)")
            }
        }`

But I'm getting an error : `error Network(The operation couldn’t be completed. Operation timed out)`

Any suggestions? Curious what @SamGabbay or @Swinny1989 ended up using (based on @Swinny1989's post it sounds like he went with SocketRocket)....

**Comment by HudsonGraeme** [comment]
I'm attempting the same thing as @flexgames - but using websocket.swift.
Let me get this clear - the full url would be 
`wss://streaming.vn.teslamotors.com/connect/(vehicle_id)`,
 the headers would look like
`["Authorization":"(email) (access_token)"]`
and one would get some data by doing
`websocket.send("control:hello")`
I'm getting a 400 Bad Request error from the following

`        let email = "foo@bar.com" //Not actual value
        let vehid = "123456789" //Not actual value
        let headers = [
        "Authorization":"\(email) \(vehid)" // the "\(variable)" is intended
        ]
        let request = NSMutableURLRequest(url: NSURL(string:"wss://streaming.vn.teslamotors.com/connect/\(vehid)")! as URL)
        request.addValue(headers)
        let ws = WebSocket(request: request as URLRequest)
        ws.allowSelfSignedSSL = true //probably unnecessary
        let send : ()->() = {
            print("send: ")
            ws.send("control:hello")
        }
        ws.event.open = {
            print("opened")
            send()
        }
        ws.event.close = { code, reason, clean in
            print("close")
        }
        ws.event.error = { error in
            print("error \(error)")
        }
        ws.event.message = { message in
            if let text = message as? String {
                print("response: \(text)")
            }
        }`
Thanks for the help and API work!

**Comment by flexgames** [comment]
@HudsonGraeme Just an FYI how I ended up band-aid-ing this. I'm having the stream flow into a file via the Node teslams files. Then I start a tail task within swift to parse the last line of the stream as it comes in. 

`func startTailTasks() {
        self.task1 = Process()
        self.task1!.terminationHandler = self.commandTerminationHandler
        let argumentsString = \"tail -n 1 -f /Users/tornado/logs/lastoutput_tesla_1.txt\"

        self.task1!.launchPath = "/bin/sh"
        self.task1!.arguments = ["-c", argumentsString]

        let pipe = Pipe()

        task1!.standardOutput = pipe
        task1!.standardError = pipe

        pipe.fileHandleForReading.waitForDataInBackgroundAndNotify()

        task1!.launch()
}


    func commandOutputNotification(_ notification: Notification) {
        let fileHandle = notification.object as! FileHandle
        let data = fileHandle.availableData

        if data.count > 0 {
            processTelemetry(data:(String.init(data: data, encoding: String.Encoding.utf8)))
            fileHandle.waitForDataInBackgroundAndNotify()
        }
    }

    func processTelemetry(data:String?) {

        guard let incomingData = data else { return }

        // Sometimes the incomingData is > 1 line of data, so we only care about the last line
        var lineArray = [String]()
        incomingData.enumerateLines { (line, stop) -> () in
            lineArray.append(line)
        }

        guard let lastOutput = lineArray.last else { return }

        let teslaStats = lastOutput.components(separatedBy: ",")

        // need 13 fields to process
        if (teslaStats.count > 12) {
...
}`
I'd love to solve this with an all-swift solution but can't seem to get past the issue above.
-Joe
P.S. Not sure why the code comments above don't format correctly. I even tried escaping the "'s but didn't seem to help.

**Comment by flexgames** [comment]
Also, I'll probably just end up using an Obj-C socket tool and doing a -Bridging-Header.h since I can't seem to get this going 100% in Swift. If you get any further let me know.

**Comment by HudsonGraeme** [comment]
@flexgames Wow. Seems way out of my league - thanks for the help, i'll probably just use JSON for now, then.

**Comment by jonasman** [comment]
Im also trying to connect to the `stream` endpoint using swift but i run into the following issue:

Im using 2 different libraries and both fail with different error messages:

> error: InvalidResponse(HTTP/1.1 200 OK)
> websocket is disconnected: Optional("Invalid HTTP upgrade")

I debugged the libs and it seems the libraries are expecting the error code in the handshake to be `101` or `0`

Do you guys have any idea why?


If i connect to the `connect` endpoint it works fine. (the autopark one)

**Comment by jonasman** [comment]
Here's the full response:

```
HTTP/1.1 200 OK
Server: nginx
Date: Sun, 23 Apr 2017 13:19:38 GMT
Content-Type: text/event-stream
Transfer-Encoding: chunked
Connection: keep-alive
Cache-Control: no-cache`

**Comment by flexgames** [comment]
FWIW @jonasman I just reloaded an old branch of my code which had used websocket.swift and substituted in wss: (I was using ws: before). I'm getting a similar message:

`2017-04-23 06:44:30.152213-0700 TeslaWatcher[36165:989224] IMKClient Stall detected, *please Report* your user scenario attaching a spindump (or sysdiagnose) that captures the problem - (invocationInterruptXPCReply) block performed very slowly (5.04 secs).
error InvalidResponse(HTTP/1.1 200 OK)
close`

**Comment by jonasman** [comment]
Im sure Tesla has changed its websockets to HTML server evetns
https://www.w3schools.com/html/html5_serversentevents.asp

Im trying to use a lib for this protocol. I will let you know if it works.

**Comment by timdorr** [comment]
BTW, W3Schools is pretty bad in terms of quality. MDN tends to have better writeups: https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events

**Comment by jonasman** [comment]
I just went for a drive but no luck. There were no events from the stream. Did anyone got lucky using the stream?

**Comment by timdorr** [comment]
If this is about the streaming API, please open up a new issue. This is unrelated to Summon.

**Comment by frenchykiller** [comment]
Hey everyone! Great job on finding all this info!

Sorry to necro a relatively old issue, but I've hit a snag when trying to call the autopark commands from my custom app. The car will start up and shift into gear but then stop again without having moved.

I've got the heartbeat going on the requested frequency and sending pings every 5 seconds but I seem to be missing somethin along the way...

Here's a snippet of what I send the car and the answer I get:

`Connected to server.
Server : {"autopark":{"autopark_pause_timeout":2000,"autopark_stop_timeout":10000,"heartbeat_frequency":500},"connection_timeout":20000,"msg_type":"control:hello"}
Server : {"autopark_style":"dead_man","msg_type":"autopark:style"}
Server : {"autopark_state":"ready","msg_type":"autopark:status"}
Server : {"homelink_nearby":false,"msg_type":"homelink:status"}
Client : {"longitude":CAR_LONG,"latitude":CAR_LAT,"msg_type":"autopark:cmd_reverse"}
Server : {"cmd_type":"autopark:cmd_reverse","msg_type":"autopark:cmd_result","reason":"","result":true}
Server : {"autopark_state":"preparing","msg_type":"autopark:status"}
Server : {"autopark_state":"active_reverse","msg_type":"autopark:status"}`

The same thing happens for cmd_forward, the only time the car did anything more than start up and wait was when I accidently sent both commands almost simultaneously at which point the wheels turned left then right but it still didn't move...

Any insight into what I'm missing would be highly appreciated, I've looked through pretty much all the posts and proposed libs that I could find but can't seem to figure out where I went wrong...

**Comment by timdorr** [comment]
Obvious question: Are you able to AutoPark from that location with the app? It might be a physical problem, but just to rule that out. 

We also might need to update things. I haven't kept this part of the code up to date (and haven't yet documented it), so the command structure and that state machine might have been updated. What car software version are you on? I'm still waiting on 2018.14 myself.

**Comment by frenchykiller** [comment]
Yup autopark from the app works like a charm, and from my code the car starts up and chimes but then refuses to budge and shuts down after a few seconds...

As for the car software I'm not sure as it's not my car, I'll try and get that info!

**Comment by davidhodge** [comment]
hey @timdorr,

How does one get the following:

"vehicle_data:location -> latitude, longitude"

Was trying to see if I could use this web sockets API for location data gathering because the old streaming API appears to be broken for some vehicles now. I see you mentioned getting long / lat data back from the server but I haven't been able to get it experimentally

**Comment by frenchykiller** [comment]
I think I remember getting the `vehicle_data:location` result by simply connecting to the web socket while the car was moving. Not sure if this was a coincidence as I was playing around with auto_park and didn't specifically attempt to get the vehicle location.

**Comment by davidhodge** [comment]
@frenchykiller thanks! yeah I just discovered that and deleted my comment before I saw yours. I've now moved on to my next problem?

Any ideas on how to get live power or speed data from the socket? The Official app appears to get the speed via the socket, but I can't tell exactly how.

**Comment by dabbler** [comment]
I've reread this a bunch of times, but it's not clear to me.   Does anyone have an API that has summon working?  I just want to use any language to be able to run my own linux script to summon my car out of the garage.  I used to have @timdorr's ruby working (and had it connected through Alexa).  But I believe that no longer works, right?  Or am I just using too old an API?

**Comment by frenchykiller** [comment]
@davidhodge 
From the testing I've done, power and speed can be octained via the REST API with an HTTP GET to the /data_request/drive_state endpoint, not sure if you can get them from the web socket. Are you sure the official app doesn't just send a GET request periodically to update the info? (Though that would be pretty messy...)

@dabbler 
Yup, I have a working auto_park (summon). Everything in this thread pertaining to the web socket is still up to date.

Basically:
-> connect to the websocket
-> send autopark_cmd:(forward|reverse|abort) to move or stop the car
-> send heartbeat_app at requested interval
-> disconnect from websocket

**Comment by droideka9990** [comment]
Reading the socket is working for me still, so I've got no clue what's going on. Is it a recent update for some of you?

**Comment by dabbler** [comment]
Might it be possible to pay someone to install (via ssh) a solution that I can call from a linux script to use the API including summon/autopark?  

I've had no luck trying to install various APIs myself, php, go, ruby, nodejs, etc.

Once I have something that can be called from the command line, then I can connect it up to my own web page and Alexa and Node Red and whatever myself.  But it's just taking me too long to get the basics of summon working.

Anyone available for this?

**Comment by dabbler** [comment]
I'm trying to get the @timdorr ruby code to summon.  But I don't see any part of the API where I initiate a forward or reverse summon.  Am I missing something?

**Comment by HudsonGraeme** [comment]
Anyone investigated smart summon connections?

**Comment by timdorr** [comment]
I haven't taken a crack at it yet, but my car _just_ got it last night. I assume it's the same WebSocket-based connection.

**Comment by atx-barnes** [comment]
Is this thread still up to date about using a web socket for the summoning feature? The last comment about it working was made by @frenchykiller in 2018.

> Yup, I have a working auto_park (summon). Everything in this thread pertaining to the web socket is still up to date.
> 
> Basically:
> -> connect to the websocket
> -> send autopark_cmd:(forward|reverse|abort) to move or stop the car
> -> send heartbeat_app at requested interval
> -> disconnect from websocket

**Comment by timdorr** [comment]
Yes, they only use WebSockets for anything that streams from the car (Summon, Smart Summon, streaming data).

**Comment by vekexasia** [comment]
Hello guys, so i'm trying to implement summon via API.
I'm connecting to `wss://streaming.vn.teslamotors.com/streaming/` 

Once connected isend then`data:subscribe_oauth` hence i receive the following:

```
{"msg_type":"control:hello","connection_timeout":0}
{"msg_type":"data:update","tag":"1876472448","value":"bla,bla,bla,bla,bla"}
{"msg_type":"data:error","tag":"1876472448","value":"disconnected","error_type":"vehicle_disconnected"}

```

In the meanwhile (before the disconnected error) i send `{msg_type: 'autopark:cmd_reverse', latitude: a, longitude:b}` 

But i don't receive any message. As far as I understand i should be receiving `autopark:status` message. or `autopark:cmd_result` but i dont receive a damn thing.

Any hint?

**Comment by timdorr** [comment]
I haven't checked, but there may be changes to the protocol due to the new auth system.

**Comment by vekexasia** [comment]
@timdorr any hint on how to sniff traffic to get how the app behaves?

**Comment by pavanpkp33** [comment]
@timdorr @vekexasia  How did you connect to WSS with auth? I see only commands going out. Any documentation that you can point me to? I'm able to use the REST endpoints with auth token.

**Comment by timdorr** [comment]
You have to send [this message](https://github.com/timdorr/tesla-api/blob/master/lib/tesla_api/stream.rb#L57-L64) once you connect to the WebSocket. There is no auth for the WebSocket connection itself.

**Comment by akramaskar** [comment]
> Hello guys, so i'm trying to implement summon via API.
> I'm connecting to `wss://streaming.vn.teslamotors.com/streaming/`
> 
> Once connected isend then`data:subscribe_oauth` hence i receive the following:
> 
> ```
> {"msg_type":"control:hello","connection_timeout":0}
> {"msg_type":"data:update","tag":"1876472448","value":"bla,bla,bla,bla,bla"}
> {"msg_type":"data:error","tag":"1876472448","value":"disconnected","error_type":"vehicle_disconnected"}
> ```
> 
> In the meanwhile (before the disconnected error) i send `{msg_type: 'autopark:cmd_reverse', latitude: a, longitude:b}`
> 
> But i don't receive any message. As far as I understand i should be receiving `autopark:status` message. or `autopark:cmd_result` but i dont receive a damn thing.
> 
> Any hint?

I'm experiencing the exact same error. Any solution on this?

**Comment by pavanpkp33** [comment]
@timdorr  - Any updates to this documentation? Will it be updated? Also, I kinda saw auth being broken again.

**Comment by joshuamking** [comment]
Is there an update to any of this? Specifically around Smart Summon?
Are there any docs for how that's supposed to work?

**Comment by Urkman** [comment]
@joshuamking Did you find any documentation for smart summon?
